cmd/watt/annotations.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/datawire/ambassador/pkg/k8s"
+	"github.com/datawire/ambassador/pkg/watt"
+)
+
+func parseAnnotationResources(resource k8s.Resource) (annotationResources []k8s.Resource, annotationErrs []watt.Error) {
+	annotationStr, annotationStrOK := resource.Metadata().Annotations()["getambassador.io/config"].(string)
+	if !annotationStrOK {
+		return
+	}
+	annotationDecoder := yaml.NewDecoder(strings.NewReader(annotationStr))
+	// Parse in to a scratch _annotationResources list instead of the final annotationResources, so that we can more
+	// easily prune invalid entries out before returning it.
+	var _annotationResources []k8s.Resource
+	for {
+		var annotationResource k8s.Resource
+		err := annotationDecoder.Decode(&annotationResource)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			annotationErrs = append(annotationErrs, watt.NewError(
+				fmt.Sprintf("%s/%s", resource.QKind(), resource.QName()),
+				"could not read YAML in annotation getambassador.io/config"))
+			return
+		}
+		_annotationResources = append(_annotationResources, annotationResource)
+	}
+	for _, annotationResource := range _annotationResources {
+		// The Canonical API Version for our resources always starts with "getambassador.io/",
+		// but it used to always start with "ambassador/". Translate as needed for backward
+		// compatibility.
+		if apiVersion := k8s.Map(annotationResource).GetString("apiVersion"); strings.HasPrefix(apiVersion, "ambassador/") {
+			annotationResource["apiVersion"] = "getambassador.io/" + strings.TrimPrefix(apiVersion, "ambassador/")
+		}
+
+		// Make sure it's in the right API group
+		if !strings.HasPrefix(k8s.Map(annotationResource).GetString("apiVersion"), "getambassador.io/") {
+			annotationErrs = append(annotationErrs, watt.NewError(
+				fmt.Sprintf("%s/%s: annotation getambassador.io/config: %s/%s",
+					resource.QKind(), resource.QName(),
+					annotationResource.QKind(), annotationResource.QName()),
```

Comment:
`annotationResource.QName()` likely won't work here, because it hasn't yet done the un-folding of the collapsed metadata.  This check should be moved down to below that un-folding.